module04/03-Panda-Social-Network/solution.py

```diff
@@ -0,0 +1,151 @@
+
+    def __eq__(self, other):
+        if isinstance(other, Panda):
+            return self.__key() == other.__key()
+        raise NotImplemented
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __key(self):
+        return self.panda_name, self.panda_email, self.panda_gender
+
+    def validate_input(self, name, email, gender):
+        if not name or not email or not gender:
+            raise ValueError("Input not provided")
+        if not self.is_valid_email(email):
+            raise ValueError("Not valid email provided")
+
+    @staticmethod
+    def is_valid_email(email):
+        email_regex = '^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$'
+        return re.search(email_regex, email)
+
+    def name(self):
+        return self.panda_name
+
+    def email(self):
+        return self.panda_email
+
+    def gender(self):
+        return self.panda_gender
+
+    def is_male(self):
+        return self.panda_gender == "male"
+
+    def is_female(self):
+        return self.panda_gender == "female"
+
+    def add_friend(self, panda):
+        self.friend_list.append(panda)
+
+    def friends(self):
+        return self.friend_list
+
+
+class PandaSocialNetwork:
+    """ Social network that can handle: adding of new :class:`Panda` users, making pandas friends,
+    checking connection level between pandas, checking how many pandas with specified gender
+    exist in pandas network"""
+
+    def __init__(self):
+        self.network = []
+
+    def add_panda(self, panda):
+        if self.has_panda(panda):
+            raise PandaAlreadyThere()
+        self.network.append(panda)
+
+    def has_panda(self, panda):
+        return panda in self.network
+
+    def make_friends(self, panda1, panda2):
+        if self.are_friends(panda1, panda2):
+            raise PandasAlreadyFriends
+        self.add_pandas_to_network_if_absent(panda1, panda2)
+        panda1.add_friend(panda2)
+        panda2.add_friend(panda1)
+
+    def add_pandas_to_network_if_absent(self, panda1, panda2):
+        if panda1 not in self.network:
+            self.add_panda(panda1)
+        if panda2 not in self.network:
+            self.add_panda(panda2)
+
+    @staticmethod
+    def are_friends(panda1, panda2):
+        if not panda1.friends() or not panda2.friends():
+            return False
+        return panda1 in panda2.friends() and panda2 in panda1.friends()
+
+    def friends_of(self, panda):
+        if not self.has_panda(panda):
+            return False
+        return panda.friends()
+
+    def connection_level(self, panda1, panda2):
+        if not self.has_panda(panda1) or not self.has_panda(panda1):
+            return False
+        if not panda1.friends():
+            return -1
+        checked_pandas = []
+        return self.__calculate_connection_level(checked_pandas, panda1, panda2)
+
+    def __calculate_connection_level(self, checked_pandas, panda1, panda2, count=0):
```

Comment:
private method start with one `_`, not two.

also, good job on utilizing default values for arguments 👍